File: go/interface-segregation/ispViolationDemonstration.go
// Interface Segregation Principle (ISP) - VIOLATION Example (Go)
//
// This example demonstrates how forcing a Robot to implement a Human interface
// violates the Interface Segregation Principle.
//
// Problem: Robot is forced to implement biological methods that make no sense
// for a robot, leading to empty implementations or panics.

package main

import (
	"errors"
	"fmt"
)

// VIOLATION: Monolithic interface forces all implementations to have methods they don't need
type Human interface {
	Work()
	Eat() error
	Sleep() error
	Breathe() error
	Think()
	Reproduce() error
	Exercise() error
	Socialize()
	FeelEmotions() error
	Dream() error
}

// Person implementation - naturally implements all interface methods
type Person struct {
	Name    string
	Energy  int
	IsAwake bool
}

func NewPerson(name string) *Person {
	return &Person{
		Name:    name,
		Energy:  100,
		IsAwake: true,
	}
}

func (p *Person) Work() {
	if !p.IsAwake {
		fmt.Printf("%s cannot work while sleeping!\n", p.Name)
		return
	}
	fmt.Printf("%s is working productively\n", p.Name)
	p.Energy -= 20
}

func (p *Person) Eat() error {
	fmt.Printf("%s is eating delicious food\n", p.Name)
	p.Energy += 30
	return nil
}

func (p *Person) Sleep() error {
	fmt.Printf("%s is sleeping peacefully\n", p.Name)
	p.IsAwake = false
	p.Energy = 100
	return nil
}

func (p *Person) Breathe() error {
	fmt.Printf("%s is breathing fresh air\n", p.Name)
	return nil
}

func (p *Person) Think() {
	fmt.Printf("%s is thinking creatively\n", p.Name)
	p.Energy -= 5
}

func (p *Person) Reproduce() error {
	fmt.Printf("%s can participate in reproduction\n", p.Name)
	return nil
}

func (p *Person) Exercise() error {
	fmt.Printf("%s is doing physical exercise\n", p.Name)
	p.Energy -= 15
	return nil
}

func (p *Person) Socialize() {
	fmt.Printf("%s is socializing with friends\n", p.Name)
	p.Energy -= 10
}

func (p *Person) FeelEmotions() error {
	fmt.Printf("%s is experiencing complex emotions\n", p.Name)
	return nil
}

func (p *Person) Dream() error {
	if !p.IsAwake {
		fmt.Printf("%s is dreaming while sleeping\n", p.Name)
	}
	return nil
}

// VIOLATION: Robot forced to implement Human interface
// This violates ISP because Robot doesn't need most of these methods
type IndustrialRobot struct {
	Model         string
	BatteryLevel  int
	IsOperational bool
}

func NewIndustrialRobot(model string) *IndustrialRobot {
	return &IndustrialRobot{
		Model:         model,
		BatteryLevel:  100,
		IsOperational: true,
	}
}

func (r *IndustrialRobot) Work() {
	if !r.IsOperational {
		fmt.Printf("%s is not operational!\n", r.Model)
		return
	}
	fmt.Printf("%s robot is performing industrial work\n", r.Model)
	r.BatteryLevel -= 10
}

// VIOLATION: Forced to implement irrelevant methods
func (r *IndustrialRobot) Eat() error {
	// Makes no sense for a robot!
	return errors.New("robots don't eat food")
}

func (r *IndustrialRobot) Sleep() error {
	// Robots don't sleep, they power down
	return errors.New("robots don't sleep! Use PowerDown() instead")
}

func (r *IndustrialRobot) Breathe() error {
	// Robots don't breathe
	return errors.New("robots don't breathe air")
}

func (r *IndustrialRobot) Think() {
	// Different kind of thinking (computational)
	fmt.Printf("%s is processing data algorithmically\n", r.Model)
	r.BatteryLevel -= 2
}

func (r *IndustrialRobot) Reproduce() error {
	// Robots are manufactured, not reproduced
	return errors.New("robots are manufactured, not reproduced")
}

func (r *IndustrialRobot) Exercise() error {
	// Robots don't need exercise
	return errors.New("robots don't need exercise")
}

func (r *IndustrialRobot) Socialize() {
	// Basic communication only
	fmt.Printf("%s is communicating via network protocols\n", r.Model)
}

func (r *IndustrialRobot) FeelEmotions() error {
	// Robots don't feel emotions
	return errors.New("robots don't have emotions")
}

func (r *IndustrialRobot) Dream() error {
	// Robots don't dream
	return errors.New("robots don't dream")
}

// Robot-specific methods that don't fit in Human interface
func (r *IndustrialRobot) PowerDown() {
	fmt.Printf("%s is powering down systems\n", r.Model)
	r.IsOperational = false
}

func (r *IndustrialRobot) Recharge() {
	fmt.Printf("%s is recharging battery\n", r.Model)
	r.BatteryLevel = 100
}

func (r *IndustrialRobot) RunDiagnostics() {
	fmt.Printf("%s is running system diagnostics\n", r.Model)
	status := "Offline"
	if r.IsOperational {
		status = "Operational"
	}
	fmt.Printf("Battery: %d%%, Status: %s\n", r.BatteryLevel, status)
}

func (r *IndustrialRobot) ExecuteProgram(program string) {
	fmt.Printf("%s is executing program: %s\n", r.Model, program)
}

// Client code that expects Human interface
type WorkManager struct{}

// This method expects all Human interface methods to work
func (wm *WorkManager) ManageWorker(worker Human) {
	fmt.Println("Managing worker...")

	// This works for both
	worker.Work()

	// These fail for Robot!
	if err := worker.Eat(); err != nil {
		fmt.Printf("Error: %s\n", err.Error())
	}

	if err := worker.Sleep(); err != nil {
		fmt.Printf("Error: %s\n", err.Error())
	}

	worker.Think() // Different behavior for robot

	if err := worker.Exercise(); err != nil {
		fmt.Printf("Error: %s\n", err.Error())
	}

	if err := worker.FeelEmotions(); err != nil {
		fmt.Printf("Error: %s\n", err.Error())
	}
}

// This method only needs work capability but is forced to accept full Human interface
func (wm *WorkManager) AssignWork(worker Human, task string) {
	fmt.Printf("Assigning task: %s\n", task)
	// We only need Work() method, but must accept entire Human interface
	worker.Work()
}

// Expects Human biological needs
func (wm *WorkManager) ProvideCare(human Human) error {
	fmt.Println("Providing human care...")

	if err := human.Eat(); err != nil {
		return fmt.Errorf("failed to provide food: %w", err)
	}

	if err := human.Sleep(); err != nil {
		return fmt.Errorf("failed to provide rest: %w", err)
	}

	if err := human.Breathe(); err != nil {
		return fmt.Errorf("failed to provide air: %w", err)
	}

	return nil
}

// Demonstrates the ISP violation
func mainInCorrect() {
	fmt.Println("=== Interface Segregation Principle - VIOLATION (Go) ===")
	fmt.Println()

	fmt.Println("1. Creating human worker:")
	person := NewPerson("Alice")

	fmt.Println("\n2. Creating robot worker:")
	robot := NewIndustrialRobot("R2D2-Industrial")

	wm := &WorkManager{}

	fmt.Println("\n3. Managing human worker (works fine):")
	wm.ManageWorker(person)

	fmt.Println("\n4. Managing robot worker (many failures!):")
	wm.ManageWorker(robot)

	fmt.Println("\n5. Attempting to provide human care to robot:")
	if err := wm.ProvideCare(robot); err != nil {
		fmt.Printf("Failed to provide care: %s\n", err.Error())
	}

	fmt.Println("\n6. Problems with this design:")
	fmt.Println("   - Robot forced to implement methods it doesn't need")
	fmt.Println("   - Many methods return errors instead of working")
	fmt.Println("   - Clients can't rely on interface methods working")
	fmt.Println("   - Robot-specific methods don't fit Human interface")
	fmt.Println("   - Violates ISP: clients forced to depend on methods they don't use")

	// Try to access robot-specific functionality
	fmt.Println("\n7. Robot-specific operations (require type assertion):")
	// Since robot is already *IndustrialRobot, we can call methods directly
	robot.PowerDown()
	robot.Recharge()
	robot.RunDiagnostics()
	robot.ExecuteProgram("Industrial Assembly v2.1")

	fmt.Println("\n=== Conclusion ===")
	fmt.Println("The monolithic Human interface forces Robot to implement")
	fmt.Println("irrelevant methods, violating the Interface Segregation Principle.")
	fmt.Println("Clients depending on Human interface cannot rely on all methods working.")
}

File: rust/interface-segregation/isp_violation_demonstration.rs
/*
 * Interface Segregation Principle (ISP) - VIOLATION Example (Rust)
 * 
 * This example demonstrates how forcing a Robot to implement a Human trait
 * violates the Interface Segregation Principle.
 * 
 * Problem: Robot is forced to implement biological methods that make no sense
 * for a robot, leading to empty implementations or panics.
 */

use std::fmt;

// VIOLATION: Monolithic trait forces all implementations to have methods they don't need
trait Human {
    fn work(&mut self);
    fn eat(&mut self) -> Result<(), String>;
    fn sleep(&mut self) -> Result<(), String>;
    fn breathe(&mut self) -> Result<(), String>;
    fn think(&mut self);
    fn reproduce(&mut self) -> Result<(), String>;
    fn exercise(&mut self) -> Result<(), String>;
    fn socialize(&mut self);
    fn feel_emotions(&mut self) -> Result<(), String>;
    fn dream(&mut self) -> Result<(), String>;
}

// Person implementation - naturally implements all trait methods
#[derive(Debug)]
struct Person {
    name: String,
    energy: i32,
    is_awake: bool,
}

impl Person {
    fn new(name: String) -> Self {
        Person {
            name,
            energy: 100,
            is_awake: true,
        }
    }
}

impl Human for Person {
    fn work(&mut self) {
        if !self.is_awake {
            println!("{} cannot work while sleeping!", self.name);
            return;
        }
        println!("{} is working productively", self.name);
        self.energy -= 20;
    }
    
    fn eat(&mut self) -> Result<(), String> {
        println!("{} is eating delicious food", self.name);
        self.energy += 30;
        Ok(())
    }
    
    fn sleep(&mut self) -> Result<(), String> {
        println!("{} is sleeping peacefully", self.name);
        self.is_awake = false;
        self.energy = 100;
        Ok(())
    }
    
    fn breathe(&mut self) -> Result<(), String> {
        println!("{} is breathing fresh air", self.name);
        Ok(())
    }
    
    fn think(&mut self) {
        println!("{} is thinking creatively", self.name);
        self.energy -= 5;
    }
    
    fn reproduce(&mut self) -> Result<(), String> {
        println!("{} can participate in reproduction", self.name);
        Ok(())
    }
    
    fn exercise(&mut self) -> Result<(), String> {
        println!("{} is doing physical exercise", self.name);
        self.energy -= 15;
        Ok(())
    }
    
    fn socialize(&mut self) {
        println!("{} is socializing with friends", self.name);
        self.energy -= 10;
    }
    
    fn feel_emotions(&mut self) -> Result<(), String> {
        println!("{} is experiencing complex emotions", self.name);
        Ok(())
    }
    
    fn dream(&mut self) -> Result<(), String> {
        if !self.is_awake {
            println!("{} is dreaming while sleeping", self.name);
        }
        Ok(())
    }
}

// VIOLATION: Robot forced to implement Human trait
// This violates ISP because Robot doesn't need most of these methods
#[derive(Debug)]
struct IndustrialRobot {
    model: String,
    battery_level: i32,
    is_operational: bool,
}

impl IndustrialRobot {
    fn new(model: String) -> Self {
        IndustrialRobot {
            model,
            battery_level: 100,
            is_operational: true,
        }
    }
    
    // Robot-specific methods that don't fit in Human trait
    fn power_down(&mut self) {
        println!("{} is powering down systems", self.model);
        self.is_operational = false;
    }
    
    fn recharge(&mut self) {
        println!("{} is recharging battery", self.model);
        self.battery_level = 100;
    }
    
    fn run_diagnostics(&self) {
        println!("{} is running system diagnostics", self.model);
        let status = if self.is_operational { "Operational" } else { "Offline" };
        println!("Battery: {}%, Status: {}", self.battery_level, status);
    }
    
    fn execute_program(&self, program: &str) {
        println!("{} is executing program: {}", self.model, program);
    }
}

impl Human for IndustrialRobot {
    fn work(&mut self) {
        if !self.is_operational {
            println!("{} is not operational!", self.model);
            return;
        }
        println!("{} robot is performing industrial work", self.model);
        self.battery_level -= 10;
    }
    
    // VIOLATION: Forced to implement irrelevant methods
    fn eat(&mut self) -> Result<(), String> {
        // Makes no sense for a robot!
        Err("Robots don't eat food!".to_string())
    }
    
    fn sleep(&mut self) -> Result<(), String> {
        // Robots don't sleep, they power down
        Err("Robots don't sleep! Use power_down() instead".to_string())
    }
    
    fn breathe(&mut self) -> Result<(), String> {
        // Robots don't breathe
        Err("Robots don't breathe air!".to_string())
    }
    
    fn think(&mut self) {
        // Different kind of thinking (computational)
        println!("{} is processing data algorithmically", self.model);
        self.battery_level -= 2;
    }
    
    fn reproduce(&mut self) -> Result<(), String> {
        // Robots are manufactured, not reproduced
        Err("Robots are manufactured, not reproduced!".to_string())
    }
    
    fn exercise(&mut self) -> Result<(), String> {
        // Robots don't need exercise
        Err("Robots don't need exercise!".to_string())
    }
    
    fn socialize(&mut self) {
        // Basic communication only
        println!("{} is communicating via network protocols", self.model);
    }
    
    fn feel_emotions(&mut self) -> Result<(), String> {
        // Robots don't feel emotions
        Err("Robots don't have emotions!".to_string())
    }
    
    fn dream(&mut self) -> Result<(), String> {
        // Robots don't dream
        Err("Robots don't dream!".to_string())
    }
}

impl fmt::Display for Person {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "Person({})", self.name)
    }
}

impl fmt::Display for IndustrialRobot {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "IndustrialRobot({})", self.model)
    }
}

// Client code that expects Human trait
struct WorkManager;

impl WorkManager {
    // This method expects all Human trait methods to work
    fn manage_worker(worker: &mut dyn Human) {
        println!("Managing worker...");
        
        // This works for both
        worker.work();
        
        // These fail for Robot!
        if let Err(e) = worker.eat() {
            println!("Error: {}", e);
        }
        
        if let Err(e) = worker.sleep() {
            println!("Error: {}", e);
        }
        
        worker.think(); // Different behavior for robot
        
        if let Err(e) = worker.exercise() {
            println!("Error: {}", e);
        }
        
        if let Err(e) = worker.feel_emotions() {
            println!("Error: {}", e);
        }
    }
    
    // This method only needs work capability but must accept full Human trait
    fn assign_work(worker: &mut dyn Human, task: &str) {
        println!("Assigning task: {}", task);
        // We only need work() method, but must accept entire Human trait
        worker.work();
    }
    
    // Expects Human biological needs
    fn provide_care(human: &mut dyn Human) -> Result<(), String> {
        println!("Providing human care...");
        
        human.eat().map_err(|e| format!("Failed to provide food: {}", e))?;
        human.sleep().map_err(|e| format!("Failed to provide rest: {}", e))?;
        human.breathe().map_err(|e| format!("Failed to provide air: {}", e))?;
        
        Ok(())
    }
}

// Function that checks if object follows Human trait contract
fn check_human_compliance() -> bool {
    // In Rust, if a type implements a trait, it automatically follows the contract
    // This is enforced at compile time
    true
}

// Demonstrates the ISP violation
fn main() {
    println!("=== Interface Segregation Principle - VIOLATION (Rust) ===");
    println!();
    
    println!("1. Creating human worker:");
    let mut person = Person::new("Alice".to_string());
    
    println!("\n2. Creating robot worker:");
    let mut robot = IndustrialRobot::new("R2D2-Industrial".to_string());
    
    println!("\n3. Checking Human trait compliance:");
    println!("Person follows Human trait: {}", check_human_compliance());
    println!("Robot follows Human trait: {}", check_human_compliance());
    
    println!("\n4. Managing human worker (works fine):");
    WorkManager::manage_worker(&mut person);
    
    println!("\n5. Managing robot worker (many failures!):");
    WorkManager::manage_worker(&mut robot);
    
    println!("\n6. Attempting to provide human care to robot:");
    if let Err(e) = WorkManager::provide_care(&mut robot) {
        println!("Failed to provide care: {}", e);
    }
    
    println!("\n7. Problems with this design:");
    println!("   - Robot forced to implement methods it doesn't need");
    println!("   - Many methods return Err instead of working");
    println!("   - Clients can't rely on trait methods working");
    println!("   - Robot-specific methods don't fit Human trait");
    println!("   - Violates ISP: clients forced to depend on methods they don't use");
    
    println!("\n8. Robot-specific operations (not part of Human trait):");
    robot.power_down();
    robot.recharge();
    robot.run_diagnostics();
    robot.execute_program("Industrial Assembly v2.1");
    
    println!("\n=== Conclusion ===");
    println!("The monolithic Human trait forces Robot to implement");
    println!("irrelevant methods, violating the Interface Segregation Principle.");
    println!("Clients depending on Human trait cannot rely on all methods working.");
}